File: backend/app/api/routes/sources.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.database.database import get_db
from app.schemas.source import SourceCreate, SourceResponse
from app.services import source_service
from app.crawler.robots import can_crawl

router = APIRouter()


@router.get("/sources", response_model=list[SourceResponse])
def list_sources(db: Session = Depends(get_db)):
    return source_service.list_sources(db)


@router.post("/sources", response_model=SourceResponse)
def create_source(source: SourceCreate, db: Session = Depends(get_db)):
    return source_service.create_source(db, source)


@router.put("/sources/{source_id}", response_model=SourceResponse)
def update_source(source_id: int, source: SourceCreate, db: Session = Depends(get_db)):
    return source_service.update_source(db, source_id, source)


@router.patch("/sources/{source_id}/status", response_model=SourceResponse)
def update_source_status(source_id: int, enabled: bool, db: Session = Depends(get_db)):
    return source_service.update_source_status(db, source_id, enabled)

@router.delete("/sources/{source_id}")
def delete_source(source_id: int, db: Session = Depends(get_db)):
    return source_service.delete_source(db, source_id)

@router.get("/sources/{source_id}/robots")
def get_source_robots_status(
    source_id: int,
    db: Session = Depends(get_db),
):
    source = source_service.get_source(
        db,
        source_id,
    )

    allowed = can_crawl(source.base_url)

    return {
        "source_id": source.id,
        "allowed": allowed,
    }

File: frontend/src/api/sources.ts
import { apiRequest } from "./client";

import type {
  Source,
  SourceCreateRequest,
} from "../types";


export function getSources(): Promise<Source[]> {
  return apiRequest<Source[]>("/sources");
}


export function createSource(
  source: SourceCreateRequest,
): Promise<Source> {
  return apiRequest<Source>(
    "/sources",
    {
      method: "POST",
      body: JSON.stringify(source),
    },
  );
}


export function updateSourceStatus(
  sourceId: number,
  enabled: boolean,
): Promise<Source> {
  return apiRequest<Source>(
    `/sources/${sourceId}/status?enabled=${enabled}`,
    {
      method: "PATCH",
    },
  );
}


export function updateSource(
  sourceId: number,
  source: {
    name: string;
    base_url: string;
    enabled_status: boolean;
    request_delay: number;
  },
): Promise<Source> {
  return apiRequest<Source>(
    `/sources/${sourceId}`,
    {
      method: "PUT",
      body: JSON.stringify(source),
    },
  );
}


export interface SourceRobotsStatus {
  source_id: number;
  allowed: boolean;
}


export function getSourceRobotsStatus(
  sourceId: number,
): Promise<SourceRobotsStatus> {
  return apiRequest<SourceRobotsStatus>(
    `/sources/${sourceId}/robots`,
  );
}